frontend_server/lib/server.dart

```diff
@@ -0,0 +1,183 @@
+library frontend_server;
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:args/args.dart';
+import 'package:front_end/compilation_message.dart';
+
+import 'package:front_end/incremental_kernel_generator.dart';
+import 'package:front_end/compiler_options.dart';
+import 'package:front_end/kernel_generator.dart';
+import 'package:kernel/ast.dart';
+import 'package:kernel/kernel.dart';
+import 'package:kernel/target/flutter.dart';
+import 'package:kernel/target/flutter_fasta.dart';
+import 'package:kernel/target/targets.dart';
+import 'package:usage/uuid/uuid.dart';
+
+ArgParser argParser = new ArgParser(allowTrailingOptions: true)
+  ..addFlag('train',
+      help: 'Run through sample command line to produce snapshot',
+      negatable: false)
+  ..addFlag('incremental',
+      help: 'Run compiler in incremental mode', defaultsTo: false)
+  ..addOption('sdk-root',
+      help: 'Path to sdk root',
+      defaultsTo: '../../out/android_debug/flutter_patched_sdk');
+
+String usage = '''
+Usage: server [options] [input.dart]
+
+If input dart source code is provided on the command line, then the server 
+compiles it, generates dill file and exits.
+If no input dart source is provided on the command line, server waits for 
+instructions from stdin.
+
+Instructions:
+- compile <input.dart>
+- recompile <boundary-key>
+<path/to/updated/file1.dart>
+<path/to/updated/file2.dart>
+...
+<boundary-key>
+- accept
+- reject
+- quit
+
+Output:
+- result <boundary-key>
+<compiler output>
+<boundary-key> [<output.dill>]
+
+Options:
+${argParser.usage}
+''';
+
+enum State { READY_FOR_INSTRUCTION, RECOMPILE_LIST }
+
+abstract class CompilerInterface {
+  compile(String filename, ArgResults options);
+  recompileDelta();
+  acceptLastDelta();
+  rejectLastDelta();
+  invalidate(Uri uri);
+}
+
+class FrontendCompiler implements CompilerInterface {
+  IncrementalKernelGenerator ikg;
+  String filename;
+
+  @override
+  compile(String filename, ArgResults options) async {
+    var program;
+    String boundaryKey = new Uuid().generateV4();
+    print("result $boundaryKey");
+    var sdkRoot = Uri.base.resolve(options['sdk-root']);
+    final CompilerOptions compilerOptions = new CompilerOptions()
+      ..sdkRoot = sdkRoot
+      ..strongMode = false
+      ..target = new FlutterTarget(new TargetFlags())
+      ..onError = (CompilationMessage message) {
+        print(message);
+      };
+    if (options['incremental']) {
+      ikg = await IncrementalKernelGenerator.newInstance(
+          compilerOptions, Uri.base.resolve(filename));
+      var deltaProgram = await ikg.computeDelta();
+      program = deltaProgram.newProgram;
+    } else {
+      // TODO(aam): Remove linkedDependencies once platform is directly embedded
+      // into VM snapshot.
+      compilerOptions.linkedDependencies = <Uri>[
+        sdkRoot.resolve('platform.dill')
+      ];
+      program = await kernelForProgram(new Uri.file(filename), compilerOptions);
```

Comment:
note that here the input argument should be consistent with the one in the incremental generator, probably this one should change to `Uri.base.resolve`